modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskMethodServiceTracker.java

```diff
@@ -67,48 +66,90 @@ public void destroy() {
 
 	public UnsafeBiFunction
 		<Company, User, BatchEngineTaskItemWriter, ReflectiveOperationException>
-			getBatchEngineTaskItemWriterFactory(String key) {
+			getBatchEngineTaskItemWriterFactory(
+				BatchEngineTaskOperation batchEngineTaskOperation,
+				String itemClassName, String apiVersion) {
 
-		return _batchEngineTaskItemWriterFactories.get(key);
+		return _batchEngineTaskItemWriterFactories.get(
+			new FactoryKey(
+				batchEngineTaskOperation, itemClassName, apiVersion));
 	}
 
 	public Class<?> getItemClass(String itemClassName) {
-		Set<Class<?>> itemClasses = _itemClassKeysMap.keySet();
-
-		Stream<Class<?>> itemClassesStream = itemClasses.stream();
-
-		return itemClassesStream.filter(
-			itemClass -> Objects.equals(itemClass.getName(), itemClassName)
-		).findFirst(
-		).orElseThrow(
-			() -> new IllegalStateException(
-				StringBundler.concat(
-					"Class ", itemClassName, " is not registered"))
-		);
+		Map.Entry<Class<?>, AtomicInteger> entry = _itemClasses.get(
+			itemClassName);
+
+		if (entry == null) {
+			throw new IllegalStateException("Unknown class :" + itemClassName);
+		}
+
+		return entry.getKey();
 	}
 
 	private final Map
-		<String,
+		<FactoryKey,
 		 UnsafeBiFunction
 			 <Company, User, BatchEngineTaskItemWriter,
 			  ReflectiveOperationException>>
 				_batchEngineTaskItemWriterFactories = new ConcurrentHashMap<>();
-	private final Map<Class<?>, List<String>> _itemClassKeysMap =
+	private final Map<String, Map.Entry<Class<?>, AtomicInteger>> _itemClasses =
```

Comment:
hi @shuyangzhou , quick question, when you said "add thread safety", does it mean that OSGI container is capable of in parallel registering/unregistering modules and components?

Reply:
@ivicac the map is updated by a service tracker. When services are registered/unregistered, the caller thread is the one driving the call to the ServiceTracker. Anyone with access (directly or indirectly) to BundleContext has the ability to register/unregister service, so the ServiceTracker callback is very much possible to be executed concurrently.

Reply:
@shuyangzhou thx